pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -222,10 +222,15 @@ func (rc *reconciler) reconcile() {
 					klog.Warningf(attachedVolume.GenerateMsgDetailed("attacherDetacher.DetachVolume started", fmt.Sprintf("This volume is not safe to detach, but maxWaitForUnmountDuration %v expired, force detaching", rc.maxWaitForUnmountDuration)))
 				}
 			}
-			if err != nil && !exponentialbackoff.IsExponentialBackoff(err) {
-				// Ignore exponentialbackoff.IsExponentialBackoff errors, they are expected.
-				// Log all other errors.
-				klog.Errorf(attachedVolume.GenerateErrorDetailed("attacherDetacher.DetachVolume failed to start", err).Error())
+			if err != nil {
+				// Add volume back to ReportAsAttached if DetachVolume call failed so that node status updater will add it back to VolumeAttached list.
+				rc.actualStateOfWorld.AddVolumeToReportAsAttached(attachedVolume.VolumeName, attachedVolume.NodeName)
```

Comment:
I will try to add unit test in a separate PR. The current unit test of reconciler does not cover this logic, need a little bit more time for it.

Reply:
Can we add them in this PR? It will take a little more time but especially if we're going to be cherry-picking this it will be worth it. If the logic is not well-covered in the unit tests that's all the more reason to update them to make sure we haven't missed anything else.

Reply:
added the unit test

Reply:
the detach code already does this if detach fails, it adds the volume back to actual state of the world. Do we still need to do this?

Reply:
It might still have a little chance to fail before executing the operation when operation_executor calls here (cannot find plugin etc.) https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_executor.go#L670

Reply:
can you add a comment about why we are adding back the device back here and in detach call both? I think this code is quite tricky and might give us hard time understanding 6months from now.

Reply:
I updated the comment. PTAL